File: Reinforcement_Learning/Thompson_Sampling/thompson_sampling.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Thompson Sampling Algorithm Implementation
Optimized for better portability and readability
"""

# Importing necessary libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import os
import random

# Get the directory where the script is located
script_dir = os.path.dirname(os.path.abspath(__file__))
file_path = os.path.join(script_dir, "Ads_CTR_Optimisation.csv")  # Dataset file

# Checking if the dataset exists
if os.path.isfile(file_path):
    dataset = pd.read_csv(file_path)
    print("Data imported successfully!")  
else:
    raise FileNotFoundError(f"Error: File not found: {file_path}. Please make sure 'Ads_CTR_Optimisation.csv' is in the same directory.")

# Implementing Thompson Sampling Algorithm
N = 10000  # Number of rounds (users)
d = 10  # Number of ads
ads_selected = []
numbers_of_rewards_1 = [0] * d  # Number of times ad i got reward 1
numbers_of_rewards_0 = [0] * d  # Number of times ad i got reward 0
total_reward = 0

for n in range(0, N):
    ad = 0
    max_random = 0
    for i in range(0, d):
        random_beta = random.betavariate(numbers_of_rewards_1[i] + 1, numbers_of_rew

File: Reinforcement_Learning/Thompson_Sampling/thompson_sampling.R
# Thompson Sampling Algorithm in R

# Importing the dataset
dataset <- read.csv('Ads_CTR_Optimisation.csv')

# Parameters
N <- 10000  # Number of rounds
d <- 10     # Number of ads
ads_selected <- integer(0)
numbers_of_rewards_1 <- rep(0, d)  # Count of rewards = 1
numbers_of_rewards_0 <- rep(0, d)  # Count of rewards = 0
total_reward <- 0

# Thompson Sampling Algorithm
for (n in 1:N) {
  max_random <- 0
  ad <- 1  # Index starts from 1 in R
  
  for (i in 1:d) {
    random_beta <- rbeta(1, numbers_of_rewards_1[i] + 1, numbers_of_rewards_0[i] + 1)
    if (random_beta > max_random) {
      max_random <- random_beta
      ad <- i
    }
  }
  
  # Select ad and get reward
  ads_selected <- c(ads_selected, ad)
  reward <- dataset[n, ad]
  
  if (reward == 1) {
    numbers_of_rewards_1[ad] <- numbers_of_rewards_1[ad] + 1
  } else {
    numbers_of_rewards_0[ad] <- numbers_of_rewards_0[ad] + 1
  }
  
  total_reward <- total_reward + reward
}

# Visualizing the Results
hist(ads_selected, col = 'blue', main = "Histogram of Ads Selection",
     xlab = "Ad Index", ylab = "Number of Selections",
     breaks = d, border = "black")

# Print total reward
print(paste("Total Reward:", total_reward))
